src/drivers/memory.js

```diff
@@ -0,0 +1,234 @@
+        var db = self._dbInfo.db;
+
+        var iterationNumber = 1;
+        for (var key in db) {
+            if (db.hasOwnProperty(key)) {
+                var value = db[key];
+
+                if (value) {
+                    value = self._dbInfo.serializer.deserialize(value);
+                }
+
+                value = iterator(value, key, iterationNumber++);
+
+                if (value !== void(0)) {
+                    return value;
+                }
+            }
+        }
+    });
+
+    executeCallback(promise, callback);
+    return promise;
+}
+
+// Same as localStorage's key() method, except takes a callback.
+function key(n, callback) {
+    var self = this;
+    var promise = self.ready().then(function() {
+        var db = self._dbInfo.db;
+        var result = null;
+        var index = 0;
+
+        for (var key in db) {
+            if (db.hasOwnProperty(key)) {
+                if (n === index) {
+                    result = key;
+                    break;
+                }
+                index++;
+            }
+        }
+
+        return result;
+    });
+
+    executeCallback(promise, callback);
+    return promise;
+}
+
+function keys(callback) {
+    var self = this;
+    var promise = self.ready().then(function() {
+        var db = self._dbInfo.db;
+        var keys = [];
+
+        for (var key in db) {
+            if (db.hasOwnProperty(key)) {
+                keys.push(key);
+            }
+        }
+
+        return keys;
+    });
+
+    executeCallback(promise, callback);
+    return promise;
+}
+
+// Supply the number of keys in the datastore to the callback function.
+function length(callback) {
+    var self = this;
+    var promise = self.keys().then(function(keys) {
+        return keys.length;
+    });
+
+    executeCallback(promise, callback);
+    return promise;
+}
+
+// Remove an item from the store, nice and simple.
+function removeItem(key, callback) {
+    var self = this;
+
+    key = normalizeKey(key);
+
+    var promise = self.ready().then(function() {
+        var db = self._dbInfo.db;
+        if (db.hasOwnProperty(key)) {
+            delete db[key];
+        }
+    });
+
+    executeCallback(promise, callback);
+    return promise;
+}
+
+// Set a key's value and run an optional callback once the value is set.
+// Unlike Gaia's implementation, the callback function is passed the value,
+// in case you want to operate on that value only after you're sure it
+// saved, or something like that.
+function setItem(key, value, callback) {
+    var self = this;
+
+    key = normalizeKey(key);
+
+    var promise = self.ready().then(function() {
+        // Convert undefined values to null.
+        // https://github.com/mozilla/localForage/pull/42
+        if (value === undefined) {
+            value = null;
+        }
+
+        // Save the original value to pass to the callback.
+        var originalValue = value;
+
+        function serializeAsync(value) {
```

Comment:
This looks like it should be a utility function. Right now it's defined every time `setItem` is called, which seems wasteful.

Reply:
:+1: I will move this outside `setItem` for now but we should definitely refactor the serializer to return promises by default.